packages/test/test-end-to-end-tests/src/test/dataMigration.spec.ts

```diff
@@ -0,0 +1,425 @@
+		existing: boolean,
+	): Promise<ContainerRuntime> {
+		this.clientDetailsType = context.clientDetails.type;
+		const codeDetails = context.getSpecifiedCodeDetails
+			? context.getSpecifiedCodeDetails()
+			: undefined;
+		assert(codeDetails !== undefined, "get code details failed");
+
+		this._currentRuntimeFactory = codeDetails.package === "v1" ? this.v1 : this.v2;
+		if (this.clientDetailsType === detachedClientType) {
+			const detachedContext = createReadOnlyContext(context, "v2");
+			this._currentRuntimeFactory = this.v1ToV2;
+			return this.currentRuntimeFactory.preInitialize(detachedContext, existing);
+		}
+
+		return this.currentRuntimeFactory.preInitialize(context, existing);
+	}
+
+	public async instantiateFirstTime(runtime: ContainerRuntime): Promise<void> {
+		return this.currentRuntimeFactory.instantiateFirstTime(runtime);
+	}
+
+	public async instantiateFromExisting(runtime: ContainerRuntime): Promise<void> {
+		return this.currentRuntimeFactory.instantiateFromExisting(runtime);
+	}
+
+	public async hasInitialized(runtime: ContainerRuntime): Promise<void> {
+		await this.currentRuntimeFactory.hasInitialized(runtime);
+
+		if (this.clientDetailsType === summarizerClientType) {
+			this.oldRuntime = runtime;
+			const readonlyRuntime = await runtime.loadDetachedAndTransitionFn();
+			this.deferred.resolve({ summarizer: this.oldRuntime, readonlyRuntime });
+		}
+	}
+}
+
+class SharedCellToSharedMapFactory implements IChannelFactory {
+	public static readonly Type = "https://graph.microsoft.com/types/cell";
+
+	public get type(): string {
+		return SharedCellToSharedMapFactory.Type;
+	}
+
+	public get attributes(): IChannelAttributes {
+		return SharedMap.getFactory().attributes;
+	}
+
+	public async load(
+		runtime: IFluidDataStoreRuntime,
+		id: string,
+		services: IChannelServices,
+		attributes: IChannelAttributes,
+	): Promise<SharedMap> {
+		const cell = new SharedCell(id, runtime, attributes);
+		await cell.load(services);
+		const map = SharedMap.getFactory().create(runtime, id) as SharedMap;
+		const val = cell.get();
+		map.set(cellKey, val as string);
+		return map;
+	}
+
+	public create(document: IFluidDataStoreRuntime, id: string): SharedMap {
+		throw new Error("Shouldn't be making a shared cell");
+	}
+}
+
+async function logTree(tree: ISnapshotTree, storage: IDocumentStorageService) {
+	const newTree: any = {};
+	newTree.blobs = {};
+	newTree.trees = {};
+	for (const [blobName, blobId] of Object.entries(tree.blobs)) {
+		const content = await readAndParse(storage, blobId);
+		newTree.blobs[blobName] = content;
+	}
+	for (const [treeName, treeNode] of Object.entries(tree.trees)) {
+		newTree.trees[treeName] = await logTree(treeNode, storage);
+	}
+	// eslint-disable-next-line @typescript-eslint/no-unsafe-return
+	return newTree;
+}
+
+async function migrate(summarizerRuntime: ContainerRuntime, readonlyRuntime: ContainerRuntime) {
+	(readonlyRuntime as any).summarizerNode.startSummary(
```

Comment:
couple notes:
- I think we need to do something here to ensure that summarizer does not produce a summary on its own, with V2 quorum state (as that's the state of it when we get here) but with V1 content.
- In general, we should consider each object in the system to know what to do in response to various events (like quorum proposals), i.e. instead of poking at objects from outside and tell them what to do, I think we benefit a lot (where it's possible to accomplish) if objects (like summarizer) know what is expected from them in response to observing transition ops. This solves first point, but also reduces number of weird APIs we need to add to various layers to be able to pick through layers.

Reply:
I agree with the notes.